nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -878,6 +881,147 @@ impl DataStore {
+                    // * Allow reconfiguration in Failed to allow changing the
+                    //   boot disk of a failed instance and free its boot disk
+                    //   for detach.
+                    // * Allow reconfiguration in Creating because one of the
+                    //   last steps of instance creation, while the instance is
+                    //   still in Creating, is to reconfigure the instance to
+                    //   the desired boot disk.
+                    let ok_to_reconfigure_instance_states = vec![
+                        InstanceState::NoVmm,
+                        InstanceState::Failed,
+                        InstanceState::Creating,
+                    ];
+
+                    let updatable = instance_dsl::instance
+                        .filter(instance_dsl::id.eq(authz_instance.id()))
+                        .filter(
+                            instance_dsl::state
+                                .eq_any(ok_to_reconfigure_instance_states),
+                        )
+                        .filter(instance_dsl::time_deleted.is_null())
+                        .select(instance_dsl::id)
+                        .first_async::<Uuid>(&conn)
+                        .await;
+
+                    if let Err(e) = updatable {
+                        if e == diesel::NotFound {
+                            return Err(err.bail(Error::not_found_by_id(
+                                ResourceType::Instance,
+                                &authz_instance.id(),
+                            )));
+                        } else {
+                            return Err(e);
+                        }
+                    }
+
+                    if let Some(disk_id) = update.boot_disk_id.clone() {
+                        // Ensure the disk is currently attached before updating
+                        // the database.
+                        let expected_state = api::external::DiskState::Attached(
+                            authz_instance.id(),
+                        );
+
+                        let attached_disk: Option<Uuid> = disk_dsl::disk
+                            .filter(disk_dsl::id.eq(disk_id))
+                            .filter(
+                                disk_dsl::attach_instance_id
+                                    .eq(authz_instance.id()),
+                            )
+                            .filter(
+                                disk_dsl::disk_state.eq(expected_state.label()),
+                            )
+                            .select(disk_dsl::id)
+                            .first_async::<Uuid>(&conn)
+                            .await
+                            .optional()?;
+
+                        if attached_disk.is_none() {
+                            return Err(err.bail(Error::conflict(
+                                "boot disk must be attached",
+                            )));
+                        }
```

Comment:
@karencfv this is where the error comes from in the case that `boot_disk` is not mentioned in the `disks` array when creating an instance - the message is a bit terse because it's the same logic as if you create an instance and try to update its `boot_disk` to something that isn't attached.

it definitely seems like this message could be confusing if you didn't realize that disks in instance creation with `"type": "create"` also implicitly attach the created disks, but i'd like to handle this by docs on `InstanceCreate`. my hunch is that either a user would be starting there and we can answer the question ahead of time, or they might check back to the docs if they see an error. i wouldn't want to try catching and rephrasing this error specifically for instance creation, lest i accidentally mischaracterize some new error in the future :grimacing:

what do you think?

Reply:
Yeah, that makes sense. Would it be too messy to do validation check beforehand in [`sic_set_boot_disk`](https://github.com/oxidecomputer/omicron/pull/6585/files#diff-60010fd773615c3263c1efda2426fa72a872dc713f62d9a4d6989f1df2cacdc5R1036), or somewhere in `instance_create.rs`? 

I can see it might get convoluted because disk attachments can only be set by name, and we're setting boot disks by name or ID, which means we'd have to do all sorts of janky stuff to check if a boot disk is included in the list of disks on create :/

But then again a good error message could save our users time of going back and forth to the docs

Reply:
oh right! i'd left a note there to nudge myself and reviewers :sweat_smile: we'd have resolved the name by attaching the disk earlier in instance creation, or it's an ID and we should have attached a disk matching that.... earlier in instance creation. i just don't know how to get that information back out of the saga to check at the start of `sic_set_boot_disk` :D 

assuming i can figure out how to get that information together, that'll be a good way to give a more informative error for sure.

Reply:
Awesome! If it starts to become too cumbersome don't worry about it. I know there's a little bit of pressure to get this feature out. We can always open up an issue for a follow up PR to improve the error message :)

Reply:
with https://github.com/oxidecomputer/omicron/commit/8b22d56212d1e8807ae14bfa50f6abea75973400, it's no longer possible to send a `params::InstanceCreate` that would set a boot disk without attaching it, so this error shouldn't be reachable anymore (unless you detach the boot disk in a race with instance creation... at which point "boot disk has to be attached" seems appropriate :D)